### PR TITLE
Worksheet 4

### PR DESCRIPTION
**Pull request by JoHaHu**

### CODE REVIEW

**Comment by tim-scholl** [comment]
Remark: We did not include the runtimes for the different simulation scenarios in our report and just the MUP/s but as MUP/s was interpreted a bit differently, the runtimes might be helpful, too. They are included in our slides at the end, if needed. Thanks!

**Comment by manishmishra6016** [comment]
> Remark: We did not include the runtimes for the different simulation scenarios in our report and just the MUP/s but as MUP/s was interpreted a bit differently, the runtimes might be helpful, too. They are included in our slides at the end, if needed. Thanks!

On the slide 18 (contest), I don't see runtimes. It would be nice if you can share them here directly.

**Comment by tim-scholl** [comment]
Of course! You're right, we only have some times (TR small and large and the collision from WS3), they are on slide 16 though.

For the contest (1000 iterations of the large Taylor-Rayleigh experiment) we achieved 00:04:250 (mm:ss:ms) on the cluster.
Locally, (with AMD Ryzen 9 7950X3D 16K/32T, 64GB RAM) we had just under one second, 00:00:931 (mm:ss:ms).

The large TR on the cluster took 08:16:46 (mm:ss:ms), while the small took 00:39:79 (mm:ss:ms).
Locally, (with AMD Ryzen 9 7950X3D 16K/32T, 64GB RAM) the respective times were 01:54:05 and 00:08:06 (mm:ss:ms).

File: src/lib/simulator/physics/Gravity.h [hunk 1]
```diff
@@ -1,15 +1,44 @@
...
+                                                                          std::array<double_v, DIMENSIONS> &corrections) {
+  SPDLOG_TRACE("Entering Gravity calculate_force_vectorized");
+
+  const auto diff = (p2.position - corrections) - p1.position;
+
+  const auto norm = ArrayUtils::L2Norm(diff);
+
+  //  if (norm == 0) {
+  //    SPDLOG_DEBUG("Zero distance between particles encountered");
+  //  }
```
Comment (manishmishra6016) [inline src/lib/simulator/physics/Gravity.h:28]: dead code to be removed

File: src/lib/simulator/physics/LennardJones.h [hunk 2]
```diff
@@ -1,19 +1,88 @@
...
+      auto sigma3 = sigma * sigma * sigma;
+      auto sigma6 = sigma3 * sigma3;
+      auto epsilon = std::sqrt(epsilons[i] * epsilons[j]);
+      auto epsilon24 = 24 * epsilon * sigma6;
+      auto epsilon48 = 2 * epsilon24 * sigma6;
+      epsilons_24[i + 8 * j] = epsilon24;
+      epsilons_48[i + 8 * j] = epsilon48;
+    }
+  }
+}
```
Comment (manishmishra6016) [inline src/lib/simulator/physics/LennardJones.h:37]: I like you are precomputing these!

File: src/lib/simulator/physics/Gravity.h [hunk 3]
```diff
@@ -1,15 +1,44 @@
...
+  SPDLOG_TRACE("Exiting Gravity calculate_force_vectorized");
+
+  auto norm_mask = norm == 0;
+  for (int i = 0; i < DIMENSIONS; ++i) {
+    stdx::where(norm_mask, force[i]) = 0;
+  }
 };
 
-}// namespace simulator::physics
+auto static calculate_force(double mass, double constant) {
```
Comment (manishmishra6016) [inline src/lib/simulator/physics/Gravity.h:40]: I wouldn't necessarily keep const gravity in the same class as force calculated from gravitational potential between particles. They are different forces. But it is fine for the worksheet.

File: src/lib/simulator/physics/Thermostat.h [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+     * @brief Applies the thermostat to adjust particle velocities.
+     * @param particles Container of particles to apply the thermostat to.
+     */
+  void inline apply(container::ParticleContainer<DIMENSIONS> &particles) {
+    double current_temperature = calculateCurrentTemperature(particles);
+    SPDLOG_DEBUG("Current temperature calculated: {}", current_temperature);
+
+    double temp_difference = t_target - current_temperature;
+    if (std::abs(temp_difference) > delta_t) {
+      temp_difference = std::signbit(temp_difference) ? -delta_t : delta_t;
```
Comment (manishmishra6016) [inline src/lib/simulator/physics/Thermostat.h:44]: you must also ensure that `delta_t` is a positive value, is that being done somewhere?  or you can simply use abs value irrespective of what user supplies

File: test/LinkedCellsTest.cpp [hunk 5]
```diff
@@ -1,219 +1,218 @@
...
+//
+//  reflecting_pc.linear([](Particle &p) {
+//    p.position = {0.5, 0.5, 0.75};
+//  });
+//
+//  reflecting_pc.boundary([](std::tuple<Particle &, Particle &> p) {
+//    auto [p1, p2] = p;
+//    EXPECT_EQ(p2.position[2], 1.25);
+//  });
+//}
```
Comment (manishmishra6016) [inline test/LinkedCellsTest.cpp:218]: Why is this whole file commented out? Are the tests moved? Fix or delete, avoid keeping dead code.

File: test/PlanetaryCalculationTest.cpp [hunk 6]
```diff
@@ -1,127 +1,127 @@
...
+//  // (L2-norm)³ = 58802662.352711186
+//
+//  const auto actual_force = 1005.34 / 58802662.352711186 * position_difference;
+//
+//  // check if each axis force is exact enough (based on the manually computed numbers)
+//  // in this case 7 digits accuracy is enough
+//  EXPECT_TRUE(calculated_force[0] - actual_force[0] < 0.000001);
+//  EXPECT_TRUE(calculated_force[1] - actual_force[1] < 0.000001);
+//  EXPECT_TRUE(calculated_force[2] - actual_force[2] < 0.000001);
+//}
```
Comment (manishmishra6016) [inline test/PlanetaryCalculationTest.cpp:127]: It is unclear as to why these tests are commented?

**Review by manishmishra6016 — APPROVED** [review]
**Note**
_This is only feedback regarding the code directly. Grade feedback will be on Moodle once we are through with all groups._

-----------------------------------------------------------------------------------------------------

Thanks for the very nice submission :muscle: 
I don't have much complaints here, everything looks nice. I was wondering why older tests were commented out, is it because of templates you introduced? Also, I have left some small comments, so you can go over them and see if they make sense.

Anyways, I really liked your work and I am pretty sure you will do great in final task as well :smile: Good luck!

**Pull request merged**